hopsandcp/dcpmaster.cpp

```diff
@@ -0,0 +1,259 @@
+#include "dcpmaster.h"
+
+#include <iostream>
+#include <cstdint>
+#include <fstream>
+#include <memory>
+
+#include "dcp/helper/LogHelper.hpp"
+#include "dcp/model/pdu/DcpPduFactory.hpp"
+#include "dcp/xml/DcpSlaveDescriptionReader.hpp"
+#include "dcp/driver/ethernet/udp/UdpDriver.hpp"
+#include "dcp/logic/DcpManagerMaster.hpp"
+
+
+DcpMaster::DcpMaster(const std::string host, u_short port)
+    : stdLog(std::cout)
+{
+    driver = new UdpDriver(host, port);
+
+    manager = new DcpManagerMaster(driver->getDcpDriver());
+
+    manager->setAckReceivedListener<SYNC>(
+                std::bind(&DcpMaster::receiveAck, this, std::placeholders::_1, std::placeholders::_2));
+    manager->setNAckReceivedListener<SYNC>(
+                std::bind(&DcpMaster::receiveNAck, this, std::placeholders::_1, std::placeholders::_2,
+                          std::placeholders::_3));
+    manager->setStateChangedNotificationReceivedListener<SYNC>(
+                std::bind(&DcpMaster::receiveStateChangedNotification, this, std::placeholders::_1,
+                          std::placeholders::_2));
+    manager->setDataReceivedListener<SYNC>(
+                std::bind(&DcpMaster::dataReceived, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+    manager->addLogListener(std::bind(&OstreamLog::logOstream, stdLog, std::placeholders::_1));
+    manager->setGenerateLogString(true);
+}
+
+DcpMaster::~DcpMaster() {
+    delete driver;
+    delete manager;
+}
+
+void DcpMaster::addSlave(string &filepath)
+{
+    slaveDescriptions.push_back(new SlaveDescription_t(*readSlaveDescription(filepath.c_str())));
+    u_char id = u_char(slaveDescriptions.size());
+    uint8_t *netInfo = new uint8_t[6];
+    //*((uint8_t *) netInfo) = *slaveDescriptions.back()->TransportProtocols.UDP_IPv4->Control->port;
+    *((uint16_t *) netInfo) = *slaveDescriptions.back()->TransportProtocols.UDP_IPv4->Control->port;
+    //*((uint8_t *) (netInfo + 2)) = asio::ip::address_v4::from_string(*slaveDescriptions.back()->TransportProtocols.UDP_IPv4->Control->host).to_uint();
+    *((uint32_t *) (netInfo + 2)) = asio::ip::address_v4::from_string(*slaveDescriptions.back()->TransportProtocols.UDP_IPv4->Control->host).to_ulong();
+    driver->getDcpDriver().setSlaveNetworkInformation(id, netInfo);
+    delete[] netInfo;
+}
+
+void DcpMaster::addConnection(size_t fromSlave, size_t fromVr, std::vector<size_t> toSlaves, std::vector<size_t> toVrs)
+{
+    std::cout << "addConnection: " << fromSlave << "," << fromVr << "," << toSlaves[0] << "," << toVrs[0] << "\n";
+    DcpConnection connection;
+    connection.fromSlave = fromSlave;
+    connection.fromVr = fromVr;
+    connection.toVrs = toVrs;
+    connection.toSlaves = toSlaves;
+    connections.push_back(connection);
+}
+
+void DcpMaster::start() {
+    std::thread b(&DcpManagerMaster::start, manager);
+    std::chrono::seconds dura(1);
+    std::this_thread::sleep_for(dura);
```

Comment:
Sleeping random time is not safe, sometimes its enough sometime not. Is there some other way to "wait long enough"

Reply:
I don't like this either but can't think of a better solution. This code snippet is taken directly from the examples provided by DCPLib.